edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/planners/KubernetesPlanner.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Planners
+{
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using k8s;
+    using Microsoft.Azure.Devices.Edge.Agent.Core;
+    using Microsoft.Azure.Devices.Edge.Agent.Docker;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Commands;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class KubernetesPlanner<T> : IPlanner
+    {
+        readonly IKubernetes client;
+        readonly ICommandFactory commandFactory;
+        readonly ICombinedConfigProvider<T> combinedConfigProvider;
+        readonly string deviceNamespace;
+        readonly string iotHubHostname;
+        readonly string deviceId;
+
+        public KubernetesPlanner(
+            string deviceNamespace,
+            string iotHubHostname,
+            string deviceId,
+            IKubernetes client,
+            ICommandFactory commandFactory,
+            ICombinedConfigProvider<T> combinedConfigProvider)
+        {
+            this.deviceNamespace = Preconditions.CheckNonWhiteSpace(deviceNamespace, nameof(deviceNamespace));
+            this.iotHubHostname = Preconditions.CheckNonWhiteSpace(iotHubHostname, nameof(iotHubHostname));
+            this.deviceId = Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+            this.commandFactory = Preconditions.CheckNotNull(commandFactory, nameof(commandFactory));
+            this.combinedConfigProvider = Preconditions.CheckNotNull(combinedConfigProvider, nameof(combinedConfigProvider));
+        }
+
+        public async Task<Plan> PlanAsync(
+            ModuleSet desired,
+            ModuleSet current,
+            IRuntimeInfo runtimeInfo,
+            IImmutableDictionary<string, IModuleIdentity> moduleIdentities)
+        {
+            Events.LogDesired(desired);
+            Events.LogCurrent(current);
+            Events.LogIdentities(moduleIdentities);
```

Comment:
For my own knowledge, is it safe to log an identity? Are we leaking a secret here?

Reply:
The log identities function below only reports the module names, so no secret is being leaked in the logs.  That would be a bad thing.